File: src/main/java/me/viniciusarnhold/altaria/command/common/InviteCommand.java
package me.viniciusarnhold.altaria.command.common;

import com.google.common.collect.ImmutableSet;
import me.viniciusarnhold.altaria.command.CommandType;
import me.viniciusarnhold.altaria.command.MessageUtils;
import me.viniciusarnhold.altaria.command.UserPermissions;
import me.viniciusarnhold.altaria.command.interfaces.ICommand;
import me.viniciusarnhold.altaria.core.BotManager;
import me.viniciusarnhold.altaria.events.utils.Commands;
import me.viniciusarnhold.altaria.utils.Actions;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import sx.blah.discord.api.events.IListener;
import sx.blah.discord.handle.impl.events.MessageReceivedEvent;
import sx.blah.discord.handle.obj.IChannel;
import sx.blah.discord.util.DiscordException;
import sx.blah.discord.util.MessageBuilder;
import sx.blah.discord.util.RateLimitException;

import java.util.EnumSet;
import java.util.List;
import java.util.Set;

/**
 * Created by Vinicius.
 *
 * @since ${PROJECT_VERSION}
 */
public class InviteCommand implements IListener<MessageReceivedEvent>, ICommand {

    private static final Logger logger = LogManager.getLogger();

    private static final String command = "Invite";

    private static final Set<String> alias = ImmutableSet.of("Inv", "Invitation");

    private static final String desc = "Returns the invite link for this bot.";

    private static final CommandType type = CommandType.BOT;

    private static final EnumSet<UserPermissions> permissions = EnumSet.noneOf(UserPermissions.class);

    /**
     * Called when the event is sent.
     *
     * @param event The event object.
     */
    @Override
    public void handle(MessageReceivedEvent event) {
        if (!isInviteCommand(event)) {
            return;
        }
        try {

            logger.traceEntry("Received Invite command with {}", () -> event.getMessage().getContent());

            List<String> args = Commands.splitByWhitespace(event.getMessage().getContent().trim());

            boolean isLocal = false;
            if (args.size() > 1 && args.get(1).equals("here")) {
                isLocal = true;
            }

            final IChannel channelToSend;
            if (!isLocal) {
                IChannel channel;
                try {
                    channel = event.getMessage().getAuthor().getOrCreatePMChannel();
                } catch (DiscordException | RateLimitException de) {
                    logger.error(de);
                    channel = event.getMessage().getChannel();
                }
                channelToSend = channel;
            } else {
                channelToSend = event.getMessage().getChannel();
            }

            MessageBuilder builder = MessageUtils.getMessageBuilder(event.getMessage())
                    .withChannel(channelToSend)
                    .appendContent("Here's my invite link!")
                    .appendContent(System.lineSeparator())
                    .appendContent(BotManager.getInstance().inviteUrl());

            MessageUtils.getDefaultRequestBuilder(event.getMessage())
                    .doAction(Actions.ofSuccess(builder::send))
                    .andThen(Actions.ofSuccess(event.getMessage()::delete))
                    .execute();

        } catch (Exception e) {
            logger.error("Failed to handle Invite command", e);
        }
    }

    private boolean isInviteCommand(MessageReceivedEvent event) {
        return !event.getMessage().getChannel().isPrivate() &&
                !event.getMessage().getAuthor().isBot() &&
                MessageUtils.isMyCommand(event.getMessage(), this);
    }

    @NotNull
    @Override
    public String command() {
        return command;
    }

    @NotNull
    @Override
    public Set<String> aliases() {
        return alias;
    }

    @NotNull
    @Override
    public String description() {
        return desc;
    }

    @NotNull
    @Override
    public CommandType type() {
        return type;
    }

    @NotNull
    @Override
    public EnumSet<UserPermissions> permissions() {
        return permissions;
    }
}

File: commands/src/main/kotlin/commands/common/InviteCommand.kt
package me.viniciusarnhold.altaria.command.common

import com.google.common.collect.ImmutableSet
import me.viniciusarnhold.altaria.command.CommandType
import me.viniciusarnhold.altaria.command.MessageUtils
import me.viniciusarnhold.altaria.command.UserPermissions
import me.viniciusarnhold.altaria.command.IMessageCommand
import me.viniciusarnhold.altaria.core.BotManager
import me.viniciusarnhold.altaria.events.utils.Commands
import me.viniciusarnhold.altaria.utils.Actions
import org.apache.logging.log4j.LogManager
import sx.blah.discord.api.events.IListener
import sx.blah.discord.handle.impl.events.MessageReceivedEvent
import sx.blah.discord.handle.obj.IChannel
import sx.blah.discord.util.DiscordException
import sx.blah.discord.util.RateLimitException
import java.util.*

/**
 * Created by Vinicius.

 * @since ${PROJECT_VERSION}
 */
class InviteCommand : IListener<MessageReceivedEvent>, IMessageCommand {

    /**
     * Called when the event is sent.

     * @param event The event object.
     */
    override fun handle(event: MessageReceivedEvent) {
        if (!isInviteCommand(event)) {
            return
        }
        try {

            logger.traceEntry("Received Invite command with {}", { event.message.content })

            val args = Commands.splitByWhitespace(event.message.content.trim { it <= ' ' })

            var isLocal = false
            if (args.size > 1 && "here" == args[1]) {
                isLocal = true
            }

            val channelToSend: IChannel
            if (!isLocal) {
                var channel: IChannel
                try {
                    channel = event.message.author.orCreatePMChannel
                } catch (de: DiscordException) {
                    logger.error(de)
                    channel = event.message.channel
                } catch (de: RateLimitException) {
                    logger.error(de)
                    channel = event.message.channel
                }

                channelToSend = channel
            } else {
                channelToSend = event.message.channel
            }

            val builder = MessageUtils.getMessageBuilder(event.message)
                    .withChannel(channelToSend)
                    .appendContent("Here's my invite link!")
                    .appendContent(System.lineSeparator())
                    .appendContent(BotManager.instance.inviteUrl())

            MessageUtils.getDefaultRequestBuilder(event.message)
                    .doAction(Actions.ofSuccess({ builder.send() }))
                    .andThen(Actions.ofSuccess({ event.message.delete() }))
                    .execute()

        } catch (e: Exception) {
            logger.error("Failed to handle Invite command", e)
        }

    }

    private fun isInviteCommand(event: MessageReceivedEvent): Boolean {
        return !event.message.channel.isPrivate &&
                !event.message.author.isBot &&
                MessageUtils.isMyCommand(event.message, this)
    }

    override fun command(): String {
        return command
    }

    override fun aliases(): Set<String> {
        return alias
    }

    override fun description(): String {
        return desc
    }

    override fun type(): me.viniciusarnhold.altaria.command.CommandType {
        return type
    }

    override fun permissions(): EnumSet<UserPermissions> {
        return permissions
    }

    companion object {

        private val logger = LogManager.getLogger()

        private val command = "Invite"

        private val alias = ImmutableSet.of("Inv", "Invitation")

        private val desc = "Returns the invite link for this bot."

        private val type = me.viniciusarnhold.altaria.command.CommandType.BOT

        private val permissions = EnumSet.noneOf<UserPermissions>(UserPermissions::class.java)
    }
}
